src/plugins/discover/public/application/main/state_management/discover_app_state_container.ts

```diff
@@ -220,17 +236,28 @@ export const getDiscoverAppStateContainer = ({
 
   const initializeAndSync = (currentSavedSearch: SavedSearch) => {
     addLog('[appState] initialize state and sync with URL', currentSavedSearch);
+
     const { data } = services;
-    const dataView = currentSavedSearch.searchSource.getField('index');
+    const savedSearchDataView = currentSavedSearch.searchSource.getField('index');
+    const appState = enhancedAppContainer.getState();
+    const setDataViewFromSavedSearch =
+      !appState.dataSource ||
+      (isDataSourceType(appState.dataSource, DataSourceType.DataView) &&
+        appState.dataSource.dataViewId !== savedSearchDataView?.id);
 
-    if (appStateContainer.getState().index !== dataView?.id) {
+    if (setDataViewFromSavedSearch) {
       // used data view is different from the given by url/state which is invalid
-      setState(appStateContainer, { index: dataView?.id });
+      setState(enhancedAppContainer, {
+        dataSource: savedSearchDataView?.id
+          ? createDataViewDataSource({ dataViewId: savedSearchDataView.id })
+          : undefined,
```

Comment:
it's been a while, yes, your thinking is correct, it was used to overwrite a given index in the URL by an index given by a saved search. and we were removing the index param anyway recently, the only thing this could be used is to actually clean up a given index in the URL to undefined, when ES|QL mode is being used. However due to this effort now we have a much cleaner separation of modes, so it should be redundant for the ES|QL mode